core/src/org/pentaho/di/core/extension/ExtensionPointMap.java

```diff
@@ -90,7 +90,10 @@ public static ExtensionPointMap getInstance() {
 
   /**
    * Add the extension point plugin to the map
-   * 
+   *
+   * Warning: Do not call this method directly through ExtensionPointMap, it may lead to concurrency problems. Call
+   * PluginRegistry.getInstance().registerPlugin() instead.
+   *
    * @param extensionPointPlugin
    */
   public void addExtensionPoint( PluginInterface extensionPointPlugin ) {
```

Comment:
@pamval, this PR approach is based on the fact that PluginRegistry, in registerPlugin() and removePlugin() methods already protects the ExtensionPoint map inside it's own lock. So by adding the callExtensionPoint() in PluginRegistry, we can use the same lock and protect this access.

Yes, HashBasedTable can be accessed by someone else and we will have the same problem. Other approach that can fix this access, without going back to the solution in the first PR, is to move the pieces of code in PluginRegistry that call ExtensionPointMap from inside it's own lock, to avoid dead lock problems. And guarantee all accesses to ExtensionPointMap are protected from concurrent access by using "synchronized" or "reentrantreadwritelock".